src/systems/physics/Physics.cc

```diff
@@ -1723,209 +1723,274 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
-        }
+    auto canonicalLink =
+        _ecm.Component<components::CanonicalLink>(entity);
 
-        // Angular velocity in body-fixed frame coordinates
-        auto bodyAngVelComp =
-            _ecm.Component<components::AngularVelocity>(_entity);
-        if (bodyAngVelComp)
-        {
-          Eigen::Vector3d bodyAngVel = R_bs * frameData.angularVelocity;
-          auto state =
-              bodyAngVelComp->SetData(math::eigen3::convert(bodyAngVel),
-              this->vec3Eql) ?
-              ComponentState::PeriodicChange :
-              ComponentState::NoChange;
-          _ecm.SetChanged(_entity, components::AngularVelocity::typeId,
-              state);
-        }
+    const auto &worldPose = frameData.pose;
+    const auto parentEntity = _ecm.ParentEntity(entity);
 
-        // Acceleration in body-fixed frame coordinates
-        auto bodyLinAccelComp =
-            _ecm.Component<components::LinearAcceleration>(_entity);
-        if (bodyLinAccelComp)
-        {
-          Eigen::Vector3d bodyLinAccel = R_bs * frameData.linearAcceleration;
-          auto state =
-              bodyLinAccelComp->SetData(math::eigen3::convert(bodyLinAccel),
-              this->vec3Eql)?
-              ComponentState::PeriodicChange :
-              ComponentState::NoChange;
-          _ecm.SetChanged(_entity, components::LinearAcceleration::typeId,
-              state);
-        }
+    // update the link or top level model pose if this is the first update,
+    // or if the link pose has changed since the last update
+    // (if the link pose hasn't changed, there's no need for a pose update)
+    const auto worldPoseMath3d = ignition::math::eigen3::convert(worldPose);
+    if ((this->linkWorldPoses.find(entity) == this->linkWorldPoses.end())
+        || !this->pose3Eql(this->linkWorldPoses[entity], worldPoseMath3d))
+    {
+      // cache the updated link pose to check if the link pose has changed
+      // during the next iteration
+      this->linkWorldPoses[entity] = worldPoseMath3d;
 
-        // Angular acceleration in world frame coordinates
-        auto bodyAngAccelComp =
-            _ecm.Component<components::AngularAcceleration>(_entity);
-        if (bodyAngAccelComp)
-        {
-          Eigen::Vector3d bodyAngAccel = R_bs * frameData.angularAcceleration;
-          auto state =
-              bodyAngAccelComp->SetData(math::eigen3::convert(bodyAngAccel),
-              this->vec3Eql) ?
-              ComponentState::PeriodicChange :
-              ComponentState::NoChange;
-          _ecm.SetChanged(_entity, components::AngularAcceleration::typeId,
-              state);
-        }
-        return true;
-      });
+      if (!canonicalLink)
+      {
+        // Compute the relative pose of this link from the parent model
+        // first get the world pose of the top level model
+        auto parentModelPoseIt = modelWorldPoses.find(parentEntity);
```

Comment:
I think this comment should be updated since we are now accessing the parent model instead of the top level model.

Reply:
Removed that line in https://github.com/ignitionrobotics/ign-gazebo/pull/685/commits/78caaffb4469c6072b1cdd5315f2deb8444cf4ab